Reflection/TypeExtension.cs

```diff
@@ -48,24 +51,98 @@ public static bool isStatic(this Type type)
+                if (type.DeclaringType == null)
+                {
+                    return type.Name;
+                }
+
+                return $"{type.DeclaringType.GetReadableName()}.{type.Name}";
+            }
+
+            Type[] genericArgs = type.GenericTypeArguments;
+
+            return GetReadableNameOfGenericRecursive(type, genericArgs, genericArgs.Length);
+        }
+
+        /// <summary>
+        /// Traverses up a nested generic type's outer types to build a human readable name.
+        /// </summary>
+        /// <param name="type">The type to get a readable name for.</param>
+        /// <param name="genericArgs">
+        /// The generic arguments of the type ordered from outer to inner generic arguments.
+        /// Ex: MyType<genericArgs[0]>.MyInnerType<genericArgs[1]>
+        /// (The default order returned by <see cref="Type.GenericTypeArguments"/>
+        /// </param>
+        /// <param name="genericArgsUpperBound">
+        /// The upper bounds (exclusive) of the <see cref="genericArgs"/> array to evaluate
+        /// </param>
+        /// <returns></returns>
+        /// <remarks>
+        /// For closed generic types the initial type is the only one with the concrete type args. This is why we pass
+        /// <see cref="genericArgs"/> through each recursion up through the outer types.
+        /// Since we recurse from inner to outer type the <see cref="genericArgsUpperBound"/> tracks the position in
+        /// the array that the current recursion should look to the left of for its generic args.
+        /// </remarks>
+        private static string GetReadableNameOfGenericRecursive(Type type, Type[] genericArgs, int genericArgsUpperBound)
+        {
+            int shallowGenericArgCount = type.GetShallowGenericTypeCount();
+            int genericArgsLowerBound = genericArgsUpperBound - shallowGenericArgCount;
+
+            Debug.Assert(genericArgsUpperBound <= genericArgs.Length);
+            Debug.Assert(genericArgsLowerBound >= 0);
+
+            // The path through any outer types
+            string outerTypes = null;
+            if (type.IsNested)
+            {
+                outerTypes = $"{GetReadableNameOfGenericRecursive(type.DeclaringType, genericArgs, genericArgsLowerBound)}.";
+            }
+
+            // If this type has no generic arguments just use the type name
+            if (shallowGenericArgCount == 0)
+            {
+                return string.Concat(outerTypes, type.Name);
             }
 
             if (type.GetGenericTypeDefinition() == s_NullableType)
             {
-                return $"{type.GenericTypeArguments[0].GetReadableName()}?";
+                return $"{outerTypes}{genericArgs[genericArgsLowerBound].GetReadableName()}?";
             }
 
             // Remove the generic type count indicator (`n) from the type name
             // Avoid having to calculate the number of digits in the generic type count by assuming it's under 100
```

Comment:
The second line of this comment can be removed now, as it's no longer using a shortcut

Reply:
Thanks, fixed!